### PR TITLE
feature(ui-ux): generate address api integration

### LINKED ISSUE #241: `Client` - Integrate `/generate-address` to Bridge App
#### What would you like to be added:
https://github.com/WavesHQ/bridge/issues/144

Every time a user tries to do DFC to EVM, user will be able to generate a wallet address on where to send funds.

Take note, once a "transaction" has started, store the cache on localStorage. This is to avoid hitting the endpoint or generating an address multiple times. 

We will add a "clear" step once a transaction has been completed.

### PR DESCRIPTION
**Pull request by fullstackninja864**
#### Which issue(s) does this PR fixes?:

Fixes #241

#### Developer Checklist:

- [ ] Read your code changes at least once
- [ ] No console errors on web
- [ ] Tested on Light mode and Dark mode\*
- [ ] Your UI implementation visually matched the rendered design\*
- [ ] Unit tests\*
- [ ] Added e2e tests\*
- [ ] Added translations\*

### CODE REVIEW

File: apps/server/src/defichain/services/WhaleWalletService.ts [hunk 1]
```diff
@@ -1,15 +1,34 @@
...
 @Injectable()
 export class WhaleWalletService {
   constructor(private readonly whaleWalletProvider: WhaleWalletProvider) {}
 
-  async generateAddress(network: SupportedNetwork = SupportedNetwork.mainnet): Promise<string> {
-    const wallet = this.whaleWalletProvider.createWallet(network);
-    return wallet.getAddress().catch((error: Error) => {
+  async generateAddress(network: SupportedNetwork = SupportedNetwork.mainnet): Promise<{ address: string }> {
+    try {
+      const lastIndex = await Prisma.pathIndex.findFirst({
```
Comment (weiyuan95) [inline apps/server/src/defichain/services/WhaleWalletService.ts:13]: What are the concurrency guarantees here? Could two servers act on the same index?

File: apps/server/test-i9n/defichain/wallet.i9n.ts [hunk 2]
```diff
@@ -1,16 +1,31 @@
...
   const WALLET_ENDPOINT = `/defichain/wallet/`;
 
   beforeAll(async () => {
+    postgreSqlContainer = await container
+      .withDatabase('bridge')
+      .withUsername('playground')
+      .withPassword('playground')
+      .withExposedPorts({
+        container: 5432,
+        host: 5432,
```
Comment (weiyuan95) [inline apps/server/test-i9n/defichain/wallet.i9n.ts:24]: Server  tests should not need to depend on a local `.env` file to run the tests. This makes the tests very brittle since we can only have 1 instance of a Postgres container running on `5432` at any given time.

I've already spoken to @greggtan on how we can handle this (since his PR is blocked by this), he will be working on the necessary changes to shift away from this design to a more dynamic one.

File: apps/server/src/defichain/services/WhaleWalletService.ts (hunk 1 above)
Comment (fullstackninja864) [reply apps/server/src/defichain/services/WhaleWalletService.ts:13]: two servers cant act on the same index with the above logic as we have unique index on `index` attribute which gives us pathIndex for address generation.
in case of concurrent request, only one of them will get fulfilled and will return the response and all remaining will get failed. on client side also we have added max retry as 3, so that in case of this concurrent failure, FE will reinitiate request and get the address

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx [hunk 3]
```diff
@@ -86,15 +90,37 @@ export default function StepTwoSendConfirmation({
...
+      } else {
+        generateAddress({ network: networkEnv })
+          .unwrap()
+          .then((data) => {
+            setStorageItem<string>(STORAGE_DFC_ADDR_KEY, data.address);
+            setDfcUniqueAddress(data.address);
+            setThrottleError("");
+          })
+          .catch(({ data }) => {
+            if (data?.statusCode === 429) {
```
Comment (pierregee) [inline apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:107]: can use enum here

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx [hunk 4]
```diff
@@ -86,15 +90,37 @@ export default function StepTwoSendConfirmation({
...
+        setDfcUniqueAddress(localDfcAddress);
+      } else {
+        generateAddress({ network: networkEnv })
+          .unwrap()
+          .then((data) => {
+            setStorageItem<string>(STORAGE_DFC_ADDR_KEY, data.address);
+            setDfcUniqueAddress(data.address);
+            setThrottleError("");
+          })
+          .catch(({ data }) => {
```
Comment (pierregee) [inline apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:106]: Can we also handle the generic error (500) for unexpected errors?
image

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx [hunk 5]
```diff
@@ -112,78 +138,91 @@ export default function StepTwoSendConfirmation({
...
+              <AddressError
+                error="Address has expired and is now unavailable for use."
+                onClick={async () => handleGenerateNewDfcAddress()}
+              />
+            ) : (
+              <div>
+                {hasTimeElapsed ? (
+                  <AddressError
+                    error="Address has expired and is now unavailable for use."
+                    onClick={async () => handleGenerateNewDfcAddress()}
```
Comment (pierregee) [inline apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:158]: Duplicate code

**Review by pierregee — CHANGES_REQUESTED** [review]

File: apps/server/src/defichain/services/WhaleWalletService.ts [hunk 6]
```diff
@@ -1,16 +1,33 @@
...
-      const wallet = this.whaleWalletProvider.createWallet(network);
-      return await wallet.getAddress();
+      const lastIndex = await Prisma.pathIndex.findFirst({
+        where: {
+          network,
+        },
+        orderBy: [{ index: 'desc' }],
+      });
+      const index = lastIndex?.index;
+      const nextIndex = index ? index + 1 : 2;
```
[the remark this answers is not in the document]
Comment (fullstackninja864) [reply apps/server/src/defichain/services/WhaleWalletService.ts:20]: If its undefined then next index will be 2, same for value 0 (as zero is false value). it the index value cant be zero until anyone add manual entry in db

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx (hunk 4 above)
Comment (fullstackninja864) [reply apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:106]: Will add this in new address creation flow PR

File: apps/server/src/defichain/services/WhaleWalletService.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pierregee) [reply apps/server/src/defichain/services/WhaleWalletService.ts:20]: Yaa just a minor nitpic, I was wondering if we could add that code explicitly as that logic above is not "immediately obvious"

File: apps/server/test-i9n/PostgreSql.i9n.ts [hunk 7]
```diff
@@ -31,42 +32,47 @@ describe('PostgreSql container', () => {
...
       },
     ];
-    await prisma.pathIndex.createMany({ data });
-    const count = await prisma.pathIndex.count();
+    await Prisma.pathIndex.createMany({ data });
+    const count = await Prisma.pathIndex.count();
     expect(count).toStrictEqual(2);
   });
 
   it('should throw error by passing duplicate index', async () => {
```
Comment (pierregee) [inline apps/server/test-i9n/PostgreSql.i9n.ts:48]: Let's add the same test case for address (duplicate)

**Review by pierregee — APPROVED** [review]
LGTM, thanks for the quick fixes! 
As discussed, 
1. Fix for the unallocated port: 5432 in e2e is not part of this PR
2. Fix for generic error generate-address endpoint will be handled on the next PR

**Comment by lykalabrada** [comment]
Can definitely be included in the new upcoming PR, but can we fix this issue and handle passing of lowercase network from UI?
image

**Review by lykalabrada — APPROVED** [review]
As discussed, some of the flow/UI changes will be covered on a separate PR 🙏

**Review by lykalabrada — APPROVED** [review]

**Review by pierregee — APPROVED** [review]

File: apps/server/src/defichain/services/WhaleWalletService.ts (hunk 6 above)
Comment (thedoublejay) [inline apps/server/src/defichain/services/WhaleWalletService.ts:20]: Shouldn't we throw an error if index is undefined? We cannot always assume it's 2

File: apps/server/src/defichain/services/WhaleWalletService.ts [hunk 8]
```diff
@@ -1,16 +1,33 @@
...
+      const wallet = this.whaleWalletProvider.createWallet(network, nextIndex);
+      const address = await wallet.getAddress();
+      await Prisma.pathIndex.create({
+        data: {
+          index: nextIndex,
+          address,
+          network,
```
Comment (thedoublejay) [inline apps/server/src/defichain/services/WhaleWalletService.ts:27]: Remove network usage

File: apps/server/src/prisma/migrations/20230203080350_init/migration.sql [hunk 9]
```diff
@@ -0,0 +1,19 @@
+-- CreateEnum
+CREATE TYPE "Network" AS ENUM ('Local', 'Playground', 'MainNet', 'TestNet', 'DevNet');
+
+-- CreateTable
+CREATE TABLE "PathIndex" (
+    "id" BIGSERIAL NOT NULL,
+    "index" INTEGER NOT NULL,
+    "address" TEXT NOT NULL,
+    "network" "Network" NOT NULL,
```
Comment (thedoublejay) [inline apps/server/src/prisma/migrations/20230203080350_init/migration.sql:9]: same, let's remove network usage

File: pnpm-lock.yaml [hunk 10]
```diff
@@ -106,12 +106,13 @@ importers:
...
       '@netlify/plugin-lighthouse': ^4.0.7
       '@netlify/plugin-nextjs': ^4.30.3
+      '@reduxjs/toolkit': ^1.9.2
       '@tailwindcss/line-clamp': ^0.4.2
       '@tailwindcss/typography': ^0.5.9
       '@testing-library/cypress': ^8.0.7
       '@types/node': 18.11.11
       '@waveshq/standard-web': ^0.41.1
-      '@waveshq/standard-web-linter': ^0.41.1
+      '@waveshq/standard-web-linter': ^0.39.0
```
Comment (thedoublejay) [inline pnpm-lock.yaml:115]: need to revert

File: pnpm-lock.yaml [hunk 11]
```diff
@@ -132,6 +133,7 @@ importers:
       '@babel/types': 7.20.7
       '@floating-ui/react-dom': 1.2.2_biqbaboplfbrettd7655fr4n2y
       '@headlessui/react': 1.7.10_biqbaboplfbrettd7655fr4n2y
+      '@reduxjs/toolkit': 1.9.2_react@18.2.0
```
Comment (thedoublejay) [inline pnpm-lock.yaml:136]: I think no need to add this, already included on standard if i;m not mistaken

**Review by thedoublejay — CHANGES_REQUESTED** [review]

File: apps/web/package.json [hunk 12]
```diff
@@ -40,6 +40,7 @@
     "@testing-library/cypress": "^8.0.7",
     "@waveshq/standard-web-linter": "^0.41.1",
     "@types/node": "18.11.11",
+    "@waveshq/standard-web-linter":  "^0.39.0",
```
Comment (thedoublejay) [inline apps/web/package.json:43]: you added it here twice

File: apps/web/package.json [hunk 13]
```diff
@@ -9,7 +9,6 @@
     "postbuild": "next-sitemap",
     "start": "next start",
     "lint": "next lint",
-    "playground:start": "docker-compose rm -fsv && docker-compose -f ../../docker-compose.yml up",
```
Comment (thedoublejay) [inline apps/web/package.json:12]: good idea that you put it on turbo. Can you remove the one on server as well, thanks!

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx [hunk 14]
```diff
@@ -112,78 +139,82 @@ export default function StepTwoSendConfirmation({
...
-              {isMobile && (
-                <>
-                  <button
-                    type="button"
-                    className="relative flex items-center text-dark-700 active:text-dark-900 mt-2"
-                    onClick={() => handleOnCopy(dfcUniqueAddress)}
+          <div>
+            {hasTimeElapsed ? (
+              <AddressError
+                error="Address has expired and is now unavailable for use."
```
Comment (thedoublejay) [inline apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:151]: Can we ask for a better copy for this?

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx [hunk 15]
```diff
@@ -1,29 +1,31 @@
...
-  }
-
-  // TODO: Replace with real api function to generate unique DFC address
-  const address: string = Array.from(Array(42), () =>
-    Math.floor(Math.random() * 36).toString(36)
-  ).join("");
-  setStorageItem<string>(STORAGE_DFC_ADDR_KEY, address);
-  return address;
-};
+function debounce(func, wait) {
```
Comment (thedoublejay) [inline apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:18]: Isn't this available on lodash?

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx [hunk 16]
```diff
@@ -86,15 +91,37 @@ export default function StepTwoSendConfirmation({
...
-  const handleGenerateNewDfcAddress = () => {
-    const newAddress = generateDfcUniqueAddress();
-    setDfcUniqueAddress(newAddress);
+  const generateDfcUniqueAddress = useCallback(
+    debounce(async () => {
+      const localDfcAddress = getStorageItem<string>(STORAGE_DFC_ADDR_KEY);
+      if (localDfcAddress) {
+        setDfcUniqueAddress(localDfcAddress);
+      } else {
+        generateAddress({ network: networkEnv })
```
Comment (thedoublejay) [inline apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:100]: await here instead of then?

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx [hunk 17]
```diff
@@ -112,78 +139,82 @@ export default function StepTwoSendConfirmation({
...
+          <div>
+            {hasTimeElapsed ? (
+              <AddressError
+                error="Address has expired and is now unavailable for use."
+                onClick={async () => handleGenerateNewDfcAddress()}
+              />
+            ) : (
+              <>
+                <div className="w-[164px] h-[164px] bg-dark-1000 p-0.5 md:rounded">
+                  <QRCode value={dfcUniqueAddress} size={160} />
```
Comment (thedoublejay) [inline apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:157]: check if this is scannable (QR)

**Review by thedoublejay — CHANGES_REQUESTED** [review]

File: apps/server/src/defichain/services/WhaleWalletService.ts (hunk 1 above)
Comment (weiyuan95) [reply apps/server/src/defichain/services/WhaleWalletService.ts:13]: Got it - i'm guessing a second `pathIndex.create` using the same index will fail since it will violate the `unique` constraint.

Do you think it's possible to test this behaviour to check that it fails as expected? Something like `Promise.all([generateAddress(), generateAddress()])`? 🤔

File: apps/server/src/defichain/services/WhaleWalletService.ts (hunk 6 above)
Comment (fullstackninja864) [reply apps/server/src/defichain/services/WhaleWalletService.ts:20]: If its undefined i.e. there is no previous entry in DB, so in that case we have to start creating address from index 2

File: apps/web/package.json (hunk 12 above)
Comment (fullstackninja864) [reply apps/web/package.json:43]: Merge conflict i guess, will remove this

File: pnpm-lock.yaml (hunk 11 above)
Comment (fullstackninja864) [reply pnpm-lock.yaml:136]: its included in reduxjs, but we dont have reduxjs installed  as we are not using it

File: apps/web/package.json (hunk 13 above)
Comment (fullstackninja864) [reply apps/web/package.json:12]: We will need one on server side to actually strat the playground

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx (hunk 14 above)
Comment (fullstackninja864) [reply apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:151]: This will get removed in next PR

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx (hunk 15 above)
Comment (fullstackninja864) [reply apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:18]: for single use I have avoided installing new lib

File: apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx (hunk 17 above)
Comment (fullstackninja864) [reply apps/web/src/components/erc-transfer/StepTwoSendConfirmation.tsx:157]: QR is working, tested it with LW input also

File: apps/server/src/defichain/services/WhaleWalletService.ts (hunk 1 above)
Comment (fullstackninja864) [reply apps/server/src/defichain/services/WhaleWalletService.ts:12]: Will address this in next PR

File: apps/web/src/layouts/contexts/NetworkEnvironmentContext.tsx [hunk 18]
```diff
@@ -7,21 +7,21 @@ import React, {
...
 
 interface NetworkContextI {
-  networkEnv: NetworkEnvironment;
+  networkEnv: EnvironmentNetwork;
   networkEnvDisplayName: string;
-  updateNetworkEnv: (networkEnv: NetworkEnvironment) => void;
+  updateNetworkEnv: (networkEnv: EnvironmentNetwork) => void;
   resetNetworkEnv: () => void;
 }
+const NETWORK_ENV_DISPLAY_NAME: Record<EnvironmentNetwork, string> = {
```
Comment (thedoublejay) [inline apps/web/src/layouts/contexts/NetworkEnvironmentContext.tsx:18]: @fullstackninja864 Isn't this the same with the values on Enum? You can fix it on your next PR

File: apps/web/src/store/website.tsx [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+  fetchBaseQuery({
+    baseUrl: process.env.BRIDGE_API_URL || "http://localhost:5741/defichain",
+  }),
+  {
+    maxRetries: 0,
+  }
+);
+
+export const bridgeApi = createApi({
+  reducerPath: "website",
```
Comment (thedoublejay) [inline apps/web/src/store/website.tsx:13]: Why is it called website? Maybe DeFiChain wallet? or defichain?

**Review by thedoublejay — CHANGES_REQUESTED** [review]

File: apps/server/prisma/schema.prisma [hunk 20]
```diff
@@ -8,21 +8,15 @@ datasource db {
...
-  PENDING
-  COMPLETED
-}
-
 enum EthereumTransactionStatus {
   NOT_CONFIRMED
   CONFIRMED
 }
 
 model PathIndex {
```
Comment (thedoublejay) [inline apps/server/prisma/schema.prisma:16]: Is it possible to rename this? To DeFiChainAddressIndex

**Review by thedoublejay — CHANGES_REQUESTED** [review]

File: apps/web/src/store/website.tsx (hunk 19 above)
Comment (fullstackninja864) [reply apps/web/src/store/website.tsx:13]: Will update this in next pr

File: apps/server/prisma/schema.prisma (hunk 20 above)
Comment (fullstackninja864) [reply apps/server/prisma/schema.prisma:16]: Will update this in next PR
